File: content/snippets/cpp/interview_lcci_17_12_binode.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
    TreeNode* prev = nullptr;

    void inorder(TreeNode* root) {
        if (!root) return;
        inorder(root->left);
        root->left = nullptr;
        prev->right = root;
        prev = root;
        inorder(root->right);
    }

public:
    TreeNode* convertBiNode(TreeNode* root) {
        TreeNode dummy(0);
        prev = &dummy;
        inorder(root);
        return dummy.right;
    }
};

int main() {
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(5);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);
    root->right->right = new TreeNode(6);
    root->left->left->left = new TreeNode(0);

    Solution sol;
    TreeNode* head = sol.convertBiNode(root);
    while (head) {
        cout << head->val << (head->right ? ' ' : '\n');
        head = head->right;
    }
    return 0;
}

File: content/snippets/go/interview_lcci_17_12_binode.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var prev *TreeNode

func inorder(root *TreeNode) {
	if root == nil {
		return
	}
	inorder(root.Left)
	root.Left = nil
	prev.Right = root
	prev = root
	inorder(root.Right)
}

func convertBiNode(root *TreeNode) *TreeNode {
	dummy := &TreeNode{}
	prev = dummy
	inorder(root)
	return dummy.Right
}

func main() {
	root := &TreeNode{Val: 4}
	root.Left = &TreeNode{Val: 2}
	root.Right = &TreeNode{Val: 5}
	root.Left.Left = &TreeNode{Val: 1}
	root.Left.Right = &TreeNode{Val: 3}
	root.Right.Right = &TreeNode{Val: 6}
	root.Left.Left.Left = &TreeNode{Val: 0}

	head := convertBiNode(root)
	for head != nil {
		if head.Right != nil {
			fmt.Printf("%d ", head.Val)
		} else {
			fmt.Printf("%d\n", head.Val)
		}
		head = head.Right
	}
}
